[#5228] hard quota enforcement

devincowan opened the pull request:
Resolves #5228 in conjunction with https://github.com/hydroshare/hydroshare/pull/5461
<!--

Please read, and add your text at the bottom of this message.

Thanks for contributing code to HydroShare. In order to maintain code quality and expedite this process, please assist the development team by making sure the following is present in this pull request.

For more information, see https://docs.google.com/document/d/1dzxqlZW5fKNEyQSeKiSFq-SmS-VOPCva95XXkBjPExs

-->

### Pull Request Checklist: 
- [x] Positive Test Case Written by Dev

<!-- Enter steps that a QA engineer, stakeholder, or user documentation writer would follow to test the positive or "successful" case of the functionality your code provides or fixes -->

- [x] Automated Testing

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Code coverage must not decrease so new functionality or code paths added during a bug fix must have appropriate tests written. Every test must pass, including PEP8 code formatting tests. -->

- [x] Sufficient User and Developer Documentation

<!-- Please email your positive test case lbrazil@cuahsi.org, who will make the decision regarding user documentation. -->

- [x] Passing Jenkins Build

<!-- Our Jenkins Instance is set up to automatically test every commit from a pull request. Every test must pass, including PEP8 code formatting tests. -->

- [x] Peer Code review and approval

<!-- This is the process by which a peer developer on the HydroShare team will read the changeset, provide feedback, and ultimately give a formal approval to the code before it passes PR status. -->

### Positive Test Case
1. Login as a non-admin user ("user") in one window
2. Use a private window to login as an admin user ("admin")
3. As "admin", use the admin panel to modify the UserQuota for "user", putting them over the hard quota limit that is set in the QuotaMessage (which you can also view in the admin panel)
4. As "user," attempt the following actions:
  * uploading a file to a resource
  * creating a resource
  * copying a resource
  * versioning a resource
  * transferring ownership of a resource to "user"
  * ingesting a bag from the userZone
  * zipping a file without removing the original
  * unzipping a file
5. All of the actions should fail
6. Now test the Quota request actions: go to the Profile page using "user"
7. Request additional quota. See the "email" printed to logs locally (depending on the email backend that you have set in settings.py)
8. Use the link in the "email" to approve the request (or use the admin panel to find the QuotaRequest and approve it)
9. See that the UserQuota allocated value for the "user" is incremented by the amount requested
10. Verify that actions are now permitted (assuming the user is now below the hard quota allocated limit)
11. All of these actions can also be tested in the API

devincowan commented on hs_core/views/utils.py:
```diff
@@ -1288,6 +1308,9 @@ def unzip_file(user, res_id, zip_with_rel_path, bool_remove_original,
                         istorage.delete(override_tgt_path)
 
+            size = validate_resource_file_size(res_files)
+            validate_user_quota(resource.get_quota_holder(), size)
```
validation?

devincowan commented:
@pkdash I added some documentation https://github.com/hydroshare/hydroshare/blob/5228-hard-quota-enforcement/quota-info.md, this PR is ready for review. Thanks!

pkdash commented on hs_core/hydroshare/resource.py:
```diff
@@ -78,14 +80,41 @@ def get_quota_usage_from_irods(username):
+    # if quotaholder is the user, get the resource size and aggregate
+    aggregate_size = 0
+    for res in resources:
+        uname = res.getAVU("quotaUserName")
```
May be better to use `res.get_quota_holder()` ?

pkdash reviewed (changes requested):
Finished 1st round of review. Mostly looks good. Have not reviewed front-end changes. May be @Maurier can help review that.

Maurier commented:
> Finished 1st round of review. Mostly looks good. Have not reviewed front-end changes. May be @Maurier can help review that.

Ok, I'll take a look at the front-end changes.

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+                    # send email to hydroshare support
+                    # CommunityRequestEmailNotification(request=request, community_request=new_quota_request,
+                    #                                   on_event=CommunityRequestEvents.CREATED).send()
+                    return HttpResponseRedirect(reverse('my_communities'))
```
good catch thanks

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+                    # CommunityRequestEmailNotification(request=request, community_request=new_quota_request,
+                    #                                   on_event=CommunityRequestEvents.CREATED).send()
+                    return HttpResponseRedirect(reverse('my_communities'))
+                except PermissionDenied:
```
removed.

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+        else:
+            raise ValidationError("Request already redeemed.")
+    except Exception as ex:
+        messages.error(request, str(ex))
```
After clicking the link, the admin is redirected and a message shown

devincowan commented on theme/utils.py:
```diff
@@ -29,56 +35,146 @@ def get_upload_path_userprofile(instance, filename):
+    <p>Thank you,</p>
+    <p>The HydroShare Team</p>
+    '''
+    send_mail(subject="HydroShare resource metadata review completed",
```
Thanks

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+        messages.error(request, str(ex))
+    else:
+        quota_request.save()
+        messages.success(request, f"Quota {action} request successful")
```
Requesting user will be notified via signal
https://github.com/hydroshare/hydroshare/pull/5229/files#diff-399f8f10080abae07bf2c6d27e71af48fd8f61299ab9a05c0dafdebbdd809449R545

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+    else:
+        quota_request.save()
+        messages.success(request, f"Quota {action} request successful")
+    return HttpResponseRedirect(request.META.get("HTTP_REFERER", "/"))
```
we want to redirect the admin user and provide message

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+            user = request.user
+            if user.is_superuser:
+                raise ObjectDoesNotExist("Admin users don't have quota")
+            up = UserProfile.objects.get(pk=user.pk)
```
thank you

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+            else:
+                for k, v in quota_form.errors.items():
+                    messages.error(request, f"Invalid {k}: {v[0]}")
+        except PermissionDenied:
```
removed

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+            messages.error(request, f"Quota request errors: {str(ex)}.")
+
+    else:
+        quota_form = QuotaRequestForm()
```
I think this was for the redirect? Sorry, lost track of what line this comment applies to...

pkdash replied:
Correct  - the `else `block

devincowan commented on hs_core/models.py:
```diff
@@ -2280,6 +2290,22 @@ def set_public(self, value, user=None):
                 if value and settings.RUN_HYRAX_UPDATE and is_netcdf_to_public:
                     run_script_to_update_hyrax_input_files(self.short_id)
 
+    def set_published(self, value):
```
Think its use got clobbered while merging conflicts from Develop
Adding it back in here: https://github.com/hydroshare/hydroshare/blob/5228-hard-quota-enforcement/hs_core/hydroshare/resource.py#L1232

devincowan commented on theme/models.py:
```diff
@@ -271,15 +283,37 @@ class Meta:
         from hs_core.hydroshare.utils import convert_file_size_to_unit
-
-        self.used_value = convert_file_size_to_unit(size, self.unit)
+        if uz_size:
```
good point, removed the conditionals

devincowan commented on theme/models.py:
```diff
@@ -433,3 +521,65 @@ def auto_delete_file_on_change(sender, instance, **kwargs):
+        return
+
+    try:
+        qr = QuotaRequest.objects.get(pk=instance.pk)
```
will be faster, good call thanks

devincowan commented on theme/models.py:
```diff
@@ -433,3 +521,65 @@ def auto_delete_file_on_change(sender, instance, **kwargs):
+        previous = UserQuota.objects.get(id=instance.id)
+        if previous.allocated_value != instance.allocated_value:
+            # allocated_value is being updated
+            instance.grace_period_ends = None
```
I think that the instance gets saved after the `pre_save` signal
https://docs.djangoproject.com/en/3.2/ref/signals/#pre-save

devincowan commented on theme/utils.py:
```diff
@@ -29,56 +35,146 @@ def get_upload_path_userprofile(instance, filename):
+               "unit": unit,
+               "uz": uz,
+               "dz": dz,
+               "uz_percent": uzp if uzp < 100 else 100,
```
I wanted to limit to 100% and then use `percent_over` to see how much over. This was because of how display is implemented on the front end. But you're right that we could just use `dz_percent` and `uz_percent` and allow them to go over 100, and remove `percent_over` (and modify the frontend to suite). Do you think that's preferable?

devincowan commented on theme/utils.py:
```diff
@@ -29,56 +35,146 @@ def get_upload_path_userprofile(instance, filename):
+        # This would indicate that the grace period has not been set even though the user went over quota.
+        # This should not happen.
+        logger.error(f"User {uq.user.username} went over quota but grace period was not set.")
+        status = QuotaStatus.GRACE_PERIOD
```
thanks. Set the default to `INFO` and removed the `else`

devincowan commented on theme/utils.py:
```diff
@@ -29,56 +35,146 @@ def get_upload_path_userprofile(instance, filename):
+    Note that percents are in the range 0 to 100
+    """
     from theme.models import QuotaMessage
     if not QuotaMessage.objects.exists():
```
good call, updated!

devincowan commented on hs_core/hydroshare/utils.py:
```diff
@@ -785,7 +805,9 @@ def validate_user_quota(user_or_username, size):
-                if used_percent >= hard_limit or uq.remaining_grace_period == 0:
+                grace_ends = uq.grace_period_ends
+                past_grace_period = grace_ends < date.today() if grace_ends else False
+                if used_percent >= hard_limit or past_grace_period:
```
Yes. Once the user is past the soft_limit allocation, they enter grace period. While in grace period, they can continue to upload files until either:
* the grace period time runs out or
* they exceed the hard_limit

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
+                uq.save()
+        updated_quota_status = updated_quota_data["status"]
+        original_quota_status = original_quota_data["status"]
+        if original_quota_status == QuotaStatus.ENFORCEMENT and updated_quota_status == QuotaStatus.ENFORCEMENT:
```
this is because we only want to notify admin if a user is already in quota enforcement -- the infraction is that they have already been notified that they are over quota and they are continuing to upload to the userZone

devincowan replied:
I refactored this into a separate function to maybe make it more clear

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -78,14 +80,41 @@ def get_quota_usage_from_irods(username):
+    """
+    Query resources by iRODS AVU quotaholder and resource state
+    :param user: the user to get quota usage for.
+    :param flag: resource state (private, discoverable, published)
```
added, thanks

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -78,14 +80,41 @@ def get_quota_usage_from_irods(username):
+    elif flag == "public":
+        resources.filter(raccess__public=True)
+    elif flag == "private":
+        resources.filter(raccess__private=True)
```
Thanks. I used `resources.filter(raccess__public=False, raccess__discoverable=False)`

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
+    published_percent = qmsg.published_resource_percent
+    user = User.objects.get(username=username)
+    published_size = get_storage_usage(user, flag="published")
+    dz -= published_size * (1 - published_percent)
```
Yes thanks. `published_percent` is 0 - 100.

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
+                uq.grace_period_ends = None
+            uq.save()
+        else:
+            if uq.grace_period_ends and uq.grace_period_ends < today :
```
good call, moved the code for resetting the grace period

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
+                    tasks.send_user_notification_at_quota_grace_start.apply_async((user.pk))
+            elif percent >= qmsg.hard_limit_percent:
+                # reset grace period when user quota exceeds hard limit
+                uq.grace_period_ends = None
```
That's a good idea. Added notification to user in the cases of exceeding soft limit or hard limit

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -106,8 +136,61 @@ def update_quota_usage(username):
+        today = datetime.date.today()
+        if percent >= qmsg.soft_limit_percent:
+            if percent >= 100 and percent < qmsg.hard_limit_percent:
+                if not uq.grace_period_ends:
```
true. I refactored to move functionality to `UserQuota` and `QuotaRequest`
Thanks!

devincowan commented on theme/views.py:
```diff
@@ -118,16 +119,156 @@ def get_context_data(self, **kwargs):
+            )
+        except ObjectDoesNotExist:
+            raise ValidationError("No matching quota request was found.")
+        if not user.is_superuser and not user == quota_request.request_from:
```
removed

devincowan commented on hs_core/hydroshare/resource.py:
```diff
@@ -78,14 +80,41 @@ def get_quota_usage_from_irods(username):
+    return aggregate_size
 
 
 def update_quota_usage(username):
```
done

devincowan commented:
@pkdash thank you for the superb review! Believe I have addressed all of your comments with one exception:
https://github.com/hydroshare/hydroshare/pull/5229#discussion_r1511764638
I wasn't sure what this comment pertained to

pkdash commented:
@devincowan I will try to review your latest changes later today or on Monday. Thanks.

pkdash commented on theme/management/commands/update_used_storage.py:
```diff
@@ -13,6 +15,21 @@ class Command(BaseCommand):
+        count = users.count()
+        for i, u in enumerate(users, start=0):
+            i += 1
             if UserQuota.objects.filter(user=u).exists() and u.is_active and not u.is_superuser:
```
Instead of checking `u.is_active` and` u.is_superuser` here we can filter out these users in line 19.

pkdash reviewed (changes requested):
Have only one more comment that I think needs code change. Thanks for addressing all previous comments.

devincowan commented:
@pkdash thanks for the review. Believe I addressed all of your finds.
Additionally, (sorry) I made some modifications to this management command:
https://github.com/hydroshare/hydroshare/blob/5228-hard-quota-enforcement/theme/management/commands/dump_user_quota.py#L12

Maurier commented on theme/templates/accounts/profile.html:
```diff
@@ -358,6 +358,73 @@ <h5><span class="text-muted">Date: </span>{{ res.created }}</h5>
+                                                {% if quota_message %}
+                                                    <div class="alert alert-info">
+                                                        {{ quota_message }}
+                                                        Additional Quota information can be found <a href="https://help.hydroshare.org/about-hydroshare/policies/quota/" target="_blank">here.</a>
```
I think this link needs to be changed to 'Read more about Additional Quota', where 'Additional Quota' is a link and has a bold font.

See https://developer.mozilla.org/en-US/docs/Learn/HTML/Introduction_to_HTML/Creating_hyperlinks#link_best_practices

devincowan replied:
good call thanks

Maurier commented on theme/templates/accounts/profile.html:
```diff
@@ -358,6 +358,73 @@ <h5><span class="text-muted">Date: </span>{{ res.created }}</h5>
+                                                {% endif %}
+                                                {% if not quota_requests %}
+                                                    <button id="quota-request-storage" type="button" data-toggle="modal" data-target="#manage-quota-dialog" class="btn btn-primary has-space-bottom">
+                                                        <i class="fa fa-plus" aria-hidden="true"></i> Request additional storage
```
I think this button is too prominent for a function that won't be used too frequently. I think it should use the `default` button style.

devincowan replied:
sounds good!

Maurier commented on theme/migrations/0033_auto_20240201_1947.py:
```diff
@@ -0,0 +1,28 @@
+        migrations.AlterField(
+            model_name='quotamessage',
+            name='enforce_content_prepend',
+            field=models.TextField(default='You can not take further action because you have exceeded your quota. Your quota for HydroShare resources is {allocated}{unit}. You currently have resources that consume {used}{unit}, {percent}% of your quota. '),
```
I think the first part of both of these paragraphs is now redundant and can be removed, given that we now have a detailed representation of the allocation in the profile page.

devincowan replied:
Agreed that it feels redundant when one is on their profile page. However, these messages are also used for messaging in other places:
* immediately after login, on the home page (https://github.com/hydroshare/hydroshare/blob/5228-hard-quota-enforcement/theme/views.py#L785)
* in this task, which sends the quota_message in an email (https://github.com/hydroshare/hydroshare/blob/5228-hard-quota-enforcement/hs_core/tasks.py#L570)


In these cases, the redundant part of the message feels more important.
Worth mentioning these messages are stored in the db and edited in the admin panel, so not limited to what we have here in this migration.

Maurier replied:
I see, in that case perhaps we could make an exception for the profile page and have it render its own message in HTML.

Maurier commented on theme/templates/accounts/profile.html:
```diff
@@ -358,6 +358,73 @@ <h5><span class="text-muted">Date: </span>{{ res.created }}</h5>
+                                                        <span class="text-danger">You are over your allocated quota by {{ qd.percent_over|floatformat:"1" }}%</span>
+                                                    {% else %}
+                                                        <div class="progress">
+                                                            <div class="progress-bar progress-bar-success" 
```
I think we could use color thresholds as it is commonly used to represent use of allocated storage.

Perhaps:
Green from 0% to 69%
Orange from 70% to 89%
Red from 90% to 100%

Colors are listed here: https://getbootstrap.com/docs/3.3/components/#progress-alternatives

devincowan replied:
I like that idea. It's a bit tricky because there are 2 progress bars here: one for the dataZone and one for the userZone. It is their sum that determines the total percent used. It doesn't really make sense to split the allocation in the middle, because each user migh use more datazone space vs userzone space.

Example: we alocate 20GB for devin. He likes to store things in hydroshare but never uses the datazone. Let's say I upload 15GB to the dataZone, I've used of 75% of my total allocation. But does it really make sense to show `Orange` for both of these progress bars, when one has 0% and one has 75%? If devin uploads 5GB more and has used 100% total allocated with 20GB in dataZone and 0GB in userZone, does it really make sense to show both progress bars in red even though one would show 0% and still be in red?

Maurier replied:
Understood. In that case we can discard this idea.

devincowan replied:
I really like the idea thanks Mauriel. I thought about using [stacked](https://getbootstrap.com/docs/3.3/components/#progress-stacked) bars with one [striped](https://getbootstrap.com/docs/3.3/components/#progress-striped) and one not in order to differentiate. Then we could use the colors as you suggested. But HS has BS3.0.2 and striped progress bars weren't added until ~BS3.3 :(

devincowan replied:
I went ahead and _ROUGHLY_ implemented part of this idea.

Three screenshots to show the progression (blue/green = under 90% usage, orange/red = 90-100%, red = over 100%):
![image](https://github.com/hydroshare/hydroshare/assets/17934193/2b1c38e6-c65f-4445-867b-c9e613238662)
![image](https://github.com/hydroshare/hydroshare/assets/17934193/b92fa412-b82e-406a-9d3e-0f1db05d8e04)
![image](https://github.com/hydroshare/hydroshare/assets/17934193/b36f6dbe-1537-4f45-8939-b07e2b3c3ba6)

Maurier replied:
This is better, but still a little unintuitive. We would need tooltip text for each bar indicating what storage it represents.

Would it be better to use two different progress bars (one above and one below) with their own breakdown and color coding?

devincowan replied:
Current version has tooltips already.

But I see your point. I can move it to two separate bars instead of stacked.

devincowan replied:
I'm going to upgrade our BS version so that I can use some `striped` bars. Back with more info once I complete that

devincowan replied:
Okay with the addition of stripes, it looks like:
### 75% of allocated quota used in DataZone, 5% used in UserZone:
![image](https://github.com/hydroshare/hydroshare/assets/17934193/06849e08-3228-416d-a898-125c5662234d)

### 45% used in DataZone, 50% in UserZone:
![image](https://github.com/hydroshare/hydroshare/assets/17934193/b26b2560-8aa8-429e-bbde-fbb83e27ebf6)

### 25% use in both zones
![image](https://github.com/hydroshare/hydroshare/assets/17934193/4c5a78bc-ec7a-4360-abac-eac1934b11b0)

Maurier commented on theme/models.py:
```diff
@@ -293,6 +327,228 @@ def add_to_used_value(self, size):
+        fields = ["justification", "storage"]
+        labels = {
+            "justification": _("Enter justification for more quota"),
+            "storage": _("How much more storage to you require? (GB)"),
```
Typo here. Change `to` to `do`.

Maurier replied:
We are rendering this form using Django forms, which is not ideal, so I would suggest using plain HTML instead so that we can
1. Make the storage input smaller and align its text to the right (as is the standard for number inputs).
2. Display the unit next to the input by using [bootstrap ](https://getbootstrap.com/docs/3.3/components/#input-groups-basic)
![image](https://github.com/hydroshare/hydroshare/assets/2448568/deea4773-0c59-42f9-892a-0beddf447688)

devincowan replied:
Great suggestions thank you. Added.

Maurier reviewed (commented):
I've left a few suggestions for you consideration, but overall looks good.

Maurier commented on theme/templates/accounts/profile.html:
```diff
@@ -358,6 +358,73 @@ <h5><span class="text-muted">Date: </span>{{ res.created }}</h5>
+                                                                <span class="sr-only">{{ qd.dz_percent|floatformat:"1" }}% of quota is used by dataZone</span>
+                                                            </div>
+                                                        </div>
+                                                        <span>{{ qd.remaining }}{{ qd.unit }} free of {{ qd.allocated }}{{ qd.unit }}</span>
```
Also, a little bit of space between the progress bar and the text below.

devincowan replied:
thanks

devincowan commented:
@pkdash I realized that there was a race condition that needed to be handled (due to asynchronous update from quota microservices). Also realized a bug that would cause quota notifications to be emailed on _EVERY_ file delete once the user was over quota. 

Made fixes in https://github.com/hydroshare/hydroshare/pull/5229/commits/fc88b9cb3557ae79e562eae2c3540e6fd974cf57 so this should be ready for your re-review.

devincowan commented on theme/models.py:
```diff
@@ -293,6 +319,238 @@ def add_to_used_value(self, size):
+
+        Note that percents are in the range 0 to 100
+        """
+        from theme.models import QuotaMessage
```
thanks removed it

devincowan commented:
@pkdash requesting a final approval on this PR. You previously approved it at commit 6b605e341104546702ea8dc33baee8e7670eb723. Since then a few changes on the frontend and everything else contained in PR https://github.com/hydroshare/hydroshare/pull/5469 which you already reviewed.
So I believe this is less a request for review and more for a final approval
